File: Backup Assistant Tests/BackupAgent_Safety.vb
Imports BackupAssistant.Core
Imports System.Collections.ObjectModel
Imports System.IO

<TestClass()>
Public Class BackupAgent_Safety
    Implements IBackupStarter

    Private _backupAgent As BackupAgent

#Region "Interface methods"

    Friend Sub PreProcess() Implements IBackupStarter.PreProcess
        ' Do nothing. This does not apply to a test.
    End Sub

    Friend Sub PostProcess() Implements IBackupStarter.PostProcess
        ' Do nothing. This does not apply to a test.
    End Sub

    Private _destinationPath As String
    Friend ReadOnly Property DestinationPath As String Implements IBackupStarter.DestinationPath
        Get
            Return _destinationPath
        End Get
    End Property

    Friend Sub ReportProgress(progress As Integer) Implements IBackupStarter.ReportProgress
        ' Do nothing. This does not apply to a test.
    End Sub

    Friend Sub ReportStatus(status As String) Implements IBackupStarter.ReportStatus
        ' Do nothing. This does not apply to a test.
    End Sub

    Friend Sub AddToLogEntry(message As String) Implements IBackupStarter.AddToLogEntry
        ' Do nothing. This does not apply to a test.
    End Sub

    Private _sourcePath As String
    Friend ReadOnly Property SourcePath As String Implements IBackupStarter.SourcePath
        Get
            Return _sourcePath
        End Get
    End Property

    Private _filterList As List(Of String)
    Private ReadOnly Property Filters As ReadOnlyCollection(Of String) Implements IBackupStarter.Filters
        Get
            Return _filterList.AsReadOnly
        End Get
    End Property

#End Region

    <TestInitialize()>
    Public Sub TestInitialize()
        _backupAgent = New BackupAgent(Me)
    End Sub

    <TestMethod()>
    Public Sub BackupAgent_Failure_SafeGetFileInfo()
        Dim fileInfo As FileInfo = Nothing

        Try
            fileInfo = _backupAgent.SafeGetFileInfo(Guid.NewGuid.ToString)
        Catch ex As Exception
            Assert.Fail("Expected no exception to be thrown. Message: " & ex.Message)
        End Try

        Assert.IsNotNull(fileInfo, "The file info was null.")
    End Sub

    <TestMethod()>
    Public Sub BackupAgent_Failure_SafeGetFiles()
        Dim files As ReadOnlyCollection(Of String) = Nothing

        Try
            files = _backupAgent.SafeGetFiles(Guid.NewGuid.ToString)
        Catch ex As Exception
            Assert.Fail("Expected no exception to be thrown. Message: " & ex.Message)
        End Try

        Assert.IsNotNull(files, "The file list was null.")
        Assert.AreEqual(0, files.Count, "Expected an empty set.")
    End Sub

    <TestMethod()>
    Public Sub BackupAgent_Failure_SafeGetDirectories()
        Dim directories As ReadOnlyCollection(Of String) = Nothing

        Try
            directories = _backupAgent.SafeGetDirectories(Guid.NewGuid.ToString)
        Catch ex As Exception
            Assert.Fail("Expected no exception to be thrown. Message: " & ex.Message)
        End Try

        Assert.IsNotNull(directories, "The directory list was null.")
        Assert.AreEqual(0, directories.Count, "Expected an empty set.")
    End Sub

    <TestMethod()>
    Public Sub BackupAgent_Failure_SafeCopyFile()
        Try
            _backupAgent.SafeCopyFile(Guid.NewGuid.ToString, Guid.NewGuid.ToString, True)
        Catch ex As Exception
            Assert.Fail("Expected no exception to be thrown. Message: " & ex.Message)
        End Try
    End Sub

    <TestMethod()>
    Public Sub BackupAgent_Failure_SafeDeleteFile()
        Try
            _backupAgent.SafeDeleteFile(Guid.NewGuid.ToString)
        Catch ex As Exception
            Assert.Fail("Expected no exception to be thrown. Message: " & ex.Message)
        End Try
    End Sub

End Class

File: test/BackupAgent.Safety.cs
using BackupAssistant.Core;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Moq;
using System;
using System.Collections.Generic;
using System.IO.Abstractions;
using System.IO.Abstractions.TestingHelpers;

namespace BackupAssistant.Tests.BackupAgent
{
    [TestClass]
    public class Safety
    {
        private Mock<IBackupStarter> _mock;
        private Core.BackupAgent _backupAgent;

        [TestInitialize]
        public void Initialize()
        {
            _mock = new Mock<IBackupStarter>(MockBehavior.Strict);
            _mock.Setup(a => a.AddToLogEntry(It.IsAny<string>()));

            _backupAgent = new Core.BackupAgent(_mock.Object);
        }

        [TestMethod]
        public void SafeGetFileInfo()
        {
            IFileInfo fileInfo = null;

            try
            {
                fileInfo = _backupAgent.SafeGetFileInfo(null);
            }
            catch (Exception e)
            {
                Assert.Fail($"Expected no exception to be thrown. Message: {e.Message}");
            }

            Assert.IsNull(fileInfo, "The file info was not null.");
        }

        [TestMethod]
        public void SafeGetFiles()
        {
            IReadOnlyCollection<string> files = null;

            try
            {
                files = _backupAgent.SafeGetFiles(Guid.NewGuid().ToString());
            }
            catch (Exception e)
            {
                Assert.Fail($"Expected no exception to be thrown. Message: {e.Message}");
            }

            Assert.IsNotNull(files, "The file list was null.");
            Assert.AreEqual(0, files.Count, "Expected an empty set.");
        }

        [TestMethod]
        public void SafeGetDirectories()
        {
            IReadOnlyCollection<string> directories = null;

            try
            {
                directories = _backupAgent.SafeGetDirectories(Guid.NewGuid().ToString());
            }
            catch (Exception e)
            {
                Assert.Fail($"Expected no exception to be thrown. Message: {e.Message}");
            }

            Assert.IsNotNull(directories, "The directory list was null.");
            Assert.AreEqual(0, directories.Count, "Expected an empty set.");
        }

        [TestMethod]
        public void EnsureDirectoryPathExists()
        {
            MockFileSystem fileSystem = new MockFileSystem();
            fileSystem.AddDirectory(@"c:\");

            _backupAgent = new Core.BackupAgent(_mock.Object, fileSystem);

            try
            {
                _backupAgent.EnsureDirectoryPathExists(@"c:\level1\level2\file.txt");
            }
            catch (Exception e)
            {
                Assert.Fail($"Expected no exception to be thrown. Message: {e.Message}");
            }

            Assert.IsTrue(fileSystem.Directory.Exists(@"c:\level1"), "Top level directory was not created.");
            Assert.IsTrue(fileSystem.Directory.Exists(@"c:\level1\level2"), "Top level directory was not created.");
        }

        [TestMethod]
        public void SafeCopyFile()
        {
            try
            {
                _backupAgent.SafeCopyFile(Guid.NewGuid().ToString(), Guid.NewGuid().ToString(), true);
            }
            catch (Exception e)
            {
                Assert.Fail($"Expected no exception to be thrown. Message: {e.Message}");
            }
        }

        [TestMethod]
        public void SafeDeleteFile()
        {
            try
            {
                _backupAgent.SafeDeleteFile(null);
            }
            catch (Exception e)
            {
                Assert.Fail($"Expected no exception to be thrown. Message: {e.Message}");
            }
        }
    }
}
